src/Sprite/OP2BmpLoader.cpp

```diff
@@ -29,7 +31,31 @@ void OP2BmpLoader::ExtractImage(std::size_t index, const std::string& filenameOu
 		throw std::runtime_error("Image height is too large to fit in standard bitmap file format.");
 	}
 
-	BitmapFile::WriteIndexed(filenameOut, imageMeta.GetBitCount(), imageMeta.width, -static_cast<int32_t>(imageMeta.height), palette, pixelContainer);
+	BitmapFile::WriteIndexed(filenameOut, imageMeta.GetBitCount(), imageMeta.width, -static_cast<int32_t>(imageMeta.height), imageMeta.scanLineByteWidth, palette, pixelContainer);
+}
+
+std::size_t OP2BmpLoader::FrameCount(std::size_t animationIndex) const {
+	return artFile.animations[animationIndex].frames.size();
+}
+
+std::size_t OP2BmpLoader::LayerCount(std::size_t animationIndex, std::size_t frameIndex) const {
+	return artFile.animations[animationIndex].frames[frameIndex].layers.size();
+}
+
+std::vector<Color> OP2BmpLoader::CreatePalette(const ImageMeta& imageMeta) const
+{
+	std::vector<Color> palette;
+
+	if (imageMeta.GetBitCount() == 1) {
+		palette.push_back(Color{ 0, 0, 0 });
+		palette.push_back(Color{ 255, 255, 255 });
+	}
+	else {
+		palette.resize(artFile.palettes[imageMeta.paletteIndex].size());
+		std::copy(artFile.palettes[imageMeta.paletteIndex].begin(), artFile.palettes[imageMeta.paletteIndex].end(), palette.begin());
```

Comment:
I haven't looked at related code yet, but my first thought upon reading this, is why are the palettes copied, rather than using a read-only reference to an existing palette?

Related, the monochrome palette could reference a global/static `PaletteMonochromeWhiteBlack` palette.

Reply:
I added a default monochrome palette. Good idea.

I tried adding the ability to return a const reference to a palette to ArtFile.h. Code doesn't seem to want to compile:

```C++
const std::vector<Color>& ArtFile::GetPalette(std::size_t index) const 
{
	if (index >= palettes.size()) {
		throw std::runtime_error("Passed index does not correspond to a palette (too large)");
	}

	return palettes[index];
}
```

Error (active)	E0312	no suitable user-defined conversion from "const Palette" to "const std::vector<Color, std::allocator<Color>>" exists	OP2Utility	\OP2Utility\src\Sprite\ArtFile.cpp	12

Ran out of time to troubleshoot.

Thanks,
Brett

Reply:
You probably need to use `const std::vector<const Color>&` as the return type. Note the `const` within the `<>`.

Reply:
Yeah, I realized this yesterday as well. I think we should just stick to copy for now. This branch is ballooning too much. We can write an issue to decide what to do with the Palette type which will then allow taking a reference if we want.